ZuluContent/Zulu/Gumps/JsGump.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Scripts.Zulu.Packets;
+using Server.Json;
+using Server.Network;
+using static Scripts.Zulu.Utilities.ZuluUtil;
+
+namespace Server.Gumps;
+
+public record JsGumpData<T>(string Type, T Data);
+
+public class JsGump<TParams, TResponse> : Gump, INotifyCompletion, IJsonGump where TParams : class where TResponse : class
+{ 
+    private readonly string m_Type;
+    private readonly Mobile m_Player;
+    private readonly TParams m_Data;
+
+    public JsGump(string type, Mobile player, TParams data) : base(default, default)
+    {
+        m_Type = type;
+        m_Player = player;
+        m_Data = data;
+    }
+
+    public override void OnResponse(NetState sender, RelayInfo info)
+    {
+        //noop
+    }
+
+    protected virtual void SendGump()
+    {
+        m_Player.NetState.AddGump(this);
+        OutgoingZuluPackets.SendJsGumpOpen(m_Player.NetState, this, new JsGumpData<TParams>(m_Type, m_Data));
+    }
+
+    #region INotifyCompletion
+
+    protected Action Continuation;
+    public bool IsCompleted { get; private set; }
+    public virtual JsGump<TParams, TResponse> GetAwaiter() => this;
+    public TResponse GetResult() => ResponseData;
+    public TResponse ResponseData { get; private set; }
+
+    public void OnCompleted(Action continuation)
+    {
+        if (!IsCompleted && ResponseData == null)
+        {
+            Continuation = continuation;
+            SendGump();
+        }
+        else
+        {
+            continuation();
+        }
+    }
+
+    #endregion
+
+    public void OnResponse(NetState ns, string base64Json)
+    {
+        // TODO: use spans
```

Comment:
Is this needed still?

Reply:
Just a note for future performance improvements using [`Span<T>`](https://docs.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay), will keep it for now as a reminder.